{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pybedtools\n",
    "import pandas\n",
    "from sklearn.cluster import KMeans\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "class Motifs:\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    input here should be motifs in peaks already\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    -----------------\n",
    "    Attributes\n",
    "    -----------------\n",
    "    \n",
    "    bam_file: alignment file in BAM format containing reads from an ATAC-seq experiment;\n",
    "              file should be sorted and indexed.\n",
    "    \n",
    "    motifs_file: transcription factor motifs in open regions of the genome formatted as:\n",
    "                 chromosome start end motif-ID score strand (score can be e-value, p-value, doesn't matter...)\n",
    "                 tab delimited, no header, should include only motifs in peaks\n",
    "    \n",
    "    motifs: holds a dictionary with motifs from motifs_file;\n",
    "            ---\n",
    "            {TF1:[(occurence1_chrom, occurence1_start, occurence1_end, occurence1_p-value, occurence1_strand),\n",
    "                  (occurence2_chrom, occurence2_start, occurence2_end, occurence2_p-value, occurence2_strand)],\n",
    "             TF2:[(occurence1_chrom, occurence1_start, occurence1_end, occurence1_p-value, occurence1_strand)]...}        \n",
    "            ---\n",
    "            Keys: motif IDs -- will be used later to build the prior, thus I use transcription factors as IDs.\n",
    "            Values: list of occurences in the genome of the motif(s) associated with identifier.\n",
    "\n",
    "    insertion_counts: holds a dictionary with matrices of insertion counts around motifs for a particular transcription factor\n",
    "                      ---\n",
    "                      Keys: motif holders -- will be used later to build the prior \n",
    "                      Values: pandas DataFrame with insertion counts around each motif for the motif(s) associated with identifier,\n",
    "                              rownames are motif holders -- chrom,start,end,strand.\n",
    "\n",
    "    tn5_9mer_counts: holds a dictionary with matrices of 9mer counts around motifs for a particular transcription factor\n",
    "                     ---\n",
    "                     Keys: motif holders -- will be used later to build the prior\n",
    "                     Values: pandas DataFrame with 9mer counts around each motif for the motif(s) associated with identifier,\n",
    "                             rownames are motif holders -- chrom,start,end,strand.\n",
    "\n",
    "    motif_clusters_insertion_counts:\n",
    "\n",
    "    motif_clusters_9mer_counts:\n",
    "\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, bam_file, motifs_file):\n",
    "        \n",
    "        self.bam_file = bam_file\n",
    "        self.motifs_file = motifs_file \n",
    "        self.motifs = {}\n",
    "        self.insertion_counts = {}\n",
    "        self.tn5_9mer_counts = {}\n",
    "        self.motif_clusters_insertion_counts = {}\n",
    "        self.motif_clusters_9mer_counts = {}\n",
    "        \n",
    "    def read_motifs_file(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        Reads in a motifs file and saves a dictionary of motifs for each transcription factor in self.motifs \n",
    "        \"\"\"\n",
    "        \n",
    "        motifs_file = self.motifs_file\n",
    "\n",
    "        with open(motifs_file, \"r\") as motifs_handle:\n",
    "\n",
    "            for motif_occurence in motifs_handle:\n",
    "\n",
    "                motif_occurence = motif_occurence.split()\n",
    "\n",
    "                motif_id = motif_occurence[4]\n",
    "                chrom = motif_occurence[0]\n",
    "                start = motif_occurence[1]\n",
    "                end = motif_occurence[2]\n",
    "                strand = motif_occurence[5]\n",
    "                score = motif_occurence[4]\n",
    "\n",
    "                to_bed = [chrom, int(start), int(end), motif_id, score, strand]\n",
    "\n",
    "                if motif_id in self.motifs:\n",
    "                    self.motifs[motif_id].append(tuple(to_bed))\n",
    "                else:\n",
    "                    self.motifs[motif_id] = [tuple(to_bed)]\n",
    "    \n",
    "\n",
    "    @staticmethod \n",
    "    def get_insertions(bam_handle, chrom, start, end, upstream = 50, downstream = 50):\n",
    "        \n",
    "        \"\"\"\n",
    "        Returns insertion counts at a given region of the genome\n",
    "        \"\"\"\n",
    "        \n",
    "\n",
    "        # initialize counts vector at 0 for all positions\n",
    "        region_length = (end - start) + 1\n",
    "        insertion_counts = [0]*region_length\n",
    "\n",
    "        # fetch reads mapping to the region specified as input\n",
    "        reads = bam_handle.fetch(chrom, start, end)\n",
    "\n",
    "        # each read represents a potential insertion within region\n",
    "        for read in reads:\n",
    "            \n",
    "            if read.is_reverse: \n",
    "                # offset by 5 bp\n",
    "                insertion_pos = read.reference_end - 5 \n",
    "            else:\n",
    "                # offset by 4 bp\n",
    "                insertion_pos = read.reference_start + 4\n",
    "            \n",
    "            pos = insertion_pos - start \n",
    "            \n",
    "            # make sure pos is within region\n",
    "            if pos in range(0, region_length):\n",
    "                insertion_counts[pos] += 1  \n",
    "                    \n",
    "        return tuple(insertion_counts)\n",
    "        \n",
    "    @staticmethod\n",
    "    def get_tn5_9mer(insertion_counts, up_offset = 4, down_offset = 5):\n",
    "        \n",
    "        \"\"\"\n",
    "        Smooth insertion counts to Tn5 occupancy (9mer track)\n",
    "        \"\"\"\n",
    "    \n",
    "        tn5_9mer_counts = list(insertion_counts)\n",
    "\n",
    "        region = range(0, len(insertion_counts))\n",
    "\n",
    "        for pos in region:\n",
    "            for idx in range(pos - up_offset, pos + down_offset + 1):\n",
    "                if idx in region and idx != pos:\n",
    "                    tn5_9mer_counts[idx] += insertion_counts[pos]\n",
    "        \n",
    "        return tuple(tn5_9mer_counts)\n",
    "            \n",
    "    \n",
    "    def compute_scores_matrices(self, upstream, downstream):\n",
    "        \n",
    "        \"\"\"\n",
    "        Generate matrices of insertion and 9mer counts for all motifs in self.motifs\n",
    "        \"\"\"\n",
    "        \n",
    "        bam_handle = pysam.AlignmentFile(self.bam_file, \"rb\")\n",
    "        \n",
    "        for motif_id in self.motifs.keys(): ## Paralelize\n",
    "            \n",
    "            insertion_counts_mat = []\n",
    "            tn5_9mer_counts_mat = []\n",
    "            rownames_motifs = []\n",
    "            \n",
    "            motifs = pybedtools.Bedtool(self.motifs[motif_id])\n",
    "            motifs = motifs.sort()\n",
    "            \n",
    "            for motif in motifs:\n",
    "                \n",
    "                # get motif coordinates -- FIMO output is a closed interval, so add 1 to end (python)\n",
    "                chrom = motif.chrom\n",
    "                center = round(stat.median(range(motif.start, motif.end + 1)))\n",
    "                start = center - upstream\n",
    "                end = center + downstream\n",
    "                \n",
    "                insertion_counts = get_insertions(bam_handle, chrom, start, end)\n",
    "                insertion_counts_mat.append(insertion_counts)\n",
    "                \n",
    "                tn5_9mer_counts = get_tn5_9mer(insertion_counts)\n",
    "                tn5_9mer_counts_mat.append(tn5_9mer_counts)\n",
    "                \n",
    "                motif_holder = \",\".join([motif.chrom, str(motif.start), str(motif.end), motif.strand])\n",
    "                rownames_motifs.append(motif_holder)\n",
    "                \n",
    "            \n",
    "            insertion_counts_df = pandas.DataFrame(insertion_counts_mat, \n",
    "                                                   index = rownames_motifs)\n",
    "            \n",
    "            tn5_9mer_counts_df = pandas.DataFrame(tn5_9mer_counts_mat, \n",
    "                                                  index = rownames_motifs)\n",
    "            \n",
    "            self.insertion_counts[motif_id] = insertion_counts_df\n",
    "            self.tn5_9mer_counts[motif_id] = tn5_9mer_counts_df\n",
    "            \n",
    "\n",
    "    def cluster_motifs(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        Use K-Means to cluster motifs for a given transcription factor in two clusters for all TFs in self.motifs\n",
    "        \"\"\"\n",
    "        ## Paralelize, as of now I am going to do both ins and 9mer (choose one later?)\n",
    "        \n",
    "        # insertion counts\n",
    "        for motif_id in self.insertion_counts.keys():\n",
    "            \n",
    "            insertion_counts_df = self.insertion_counts[motif_id]\n",
    "            insertion_counts_mat = insertion_counts_df.as_matrix()\n",
    "\n",
    "            motif_clusters = KMeans(n_clusters = 2)\n",
    "            motif_clusters.fit(insertion_counts_mat)\n",
    "            labels = motif_clusters.labels_\n",
    "            results = pandas.DataFrame([insertion_counts_df.index, labels]).T\n",
    "            results.columns = ['motif', 'cluster']     \n",
    "            \n",
    "            clust0 = insertion_counts_df[insertion_counts_df.index.isin(results['motif'][results['cluster'] == 0])]\n",
    "            clust1 = insertion_counts_df[insertion_counts_df.index.isin(results['motif'][results['cluster'] == 1])]\n",
    "            \n",
    "            clust0_mean = clust0.mean(1).mean()\n",
    "            clust1_mean = clust1.mean(1).mean()\n",
    "            \n",
    "            if clust0_mean > clust1_mean:\n",
    "                \n",
    "                results['cluster'][results['cluster'] == 0] = 'keep'\n",
    "                results['cluster'][results['cluster'] == 1] = 'not_keep'\n",
    "            \n",
    "            if clust1_mean > clust0_mean:\n",
    "                \n",
    "                results['cluster'][results['cluster'] == 1] = 'keep'\n",
    "                results['cluster'][results['cluster'] == 0] = 'not_keep'\n",
    "            \n",
    "            self.motif_clusters_insertion_counts[motif_id] = results\n",
    "            \n",
    "            \n",
    "            \n",
    "        # 9mers\n",
    "        for motif_id in self.tn5_9mer_counts.keys():\n",
    "            \n",
    "            tn5_9mer_counts_df = self.tn5_9mer_counts[motif_id]\n",
    "            tn5_9mer_counts_mat = tn5_9mer_counts_df.as_matrix()\n",
    "\n",
    "            motif_clusters = KMeans(n_clusters = 2)\n",
    "            motif_clusters.fit(tn5_9mer_counts_mat)\n",
    "            labels = motif_clusters.labels_\n",
    "            results = pandas.DataFrame([tn5_9mer_counts_df.index, labels]).T\n",
    "            results.columns = ['motif', 'cluster']\n",
    "            \n",
    "            clust0 = insertion_counts_df[insertion_counts_df.index.isin(results['motif'][results['cluster'] == 0])]\n",
    "            clust1 = insertion_counts_df[insertion_counts_df.index.isin(results['motif'][results['cluster'] == 1])]\n",
    "            \n",
    "            clust0_mean = clust0.mean(1).mean()\n",
    "            clust1_mean = clust1.mean(1).mean()\n",
    "            \n",
    "            if clust0_mean > clust1_mean:\n",
    "                \n",
    "                results['cluster'][results['cluster'] == 0] = 'keep'\n",
    "                results['cluster'][results['cluster'] == 1] = 'not_keep'\n",
    "            \n",
    "            if clust1_mean > clust0_mean:\n",
    "                \n",
    "                results['cluster'][results['cluster'] == 1] = 'keep'\n",
    "                results['cluster'][results['cluster'] == 0] = 'not_keep'\n",
    "            \n",
    "            self.motif_clusters_9mer_counts[motif_id] = results\n",
    "\n",
    "            \n",
    "\n",
    "class Edge:\n",
    "    \n",
    "    def __init__(self, regulator, target, weight, motif, gene):\n",
    "        \n",
    "        self.regulator = regulator\n",
    "        self.target = target\n",
    "        self.weight = weight\n",
    "        self.motif = motif\n",
    "        self.gene = gene\n",
    "        \n",
    "        \n",
    "class Prior:\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    -----------------\n",
    "    Attributes\n",
    "    -----------------\n",
    "    \n",
    "    bam_file: aligned ATAC-seq BAM file\n",
    "    \n",
    "    motifs_file: motifs annotation in FIMO format IN PEAKS\n",
    "    \n",
    "    peaks_file: ATAC peaks in BED format (e.g. called by MACS)\n",
    "    \n",
    "    tss_file: TSS annotation in BED format\n",
    "    \n",
    "    transcription_factors:\n",
    "    \n",
    "    target_genes:\n",
    "    \n",
    "    upstream_in_matrix: \n",
    "    \n",
    "    downstream_in_matrix:\n",
    "    \n",
    "    cluster: whether to cluster motifs or not \n",
    "    \n",
    "    max_upstream_distance: bp upstream allowed for assignment peak to gene\n",
    "    \n",
    "    max_downstream_distance: bp downstream allowed for assignment peak to gene\n",
    "        \n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, bam_file, motifs_file,  peaks_file, tss_file,\n",
    "                 transcription_factors, target_genes, cluster = True,\n",
    "                 upstream_in_matrix = 50, downstream_in_matrix = 50, \n",
    "                 max_upstream_distance = float(\"Inf\"), max_downstream_distance = float(\"Inf\")):\n",
    "\n",
    "        self.bam_file = bam_file     \n",
    "        self.motifs_file = motifs_file \n",
    "        self.peaks_file = peaks_file   \n",
    "        self.tss_file = tss_file\n",
    "        self.transcription_factors = transcription_factors\n",
    "        self.target_genes = target_genes\n",
    "        self.upstream_in_matrix = upstream_in_matrix\n",
    "        self.downstream_in_matrix = downstream_in_matrix\n",
    "        self.cluster = cluster \n",
    "        self.max_upstream_distance = max_upstream_distance     \n",
    "        self.max_downstream_distance = max_downstream_distance \n",
    "        \n",
    "        self.motifs = []\n",
    "        self.peaks = []\n",
    "        self.tss = []\n",
    "        \n",
    "        \n",
    "    def __str__(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        print \"Prior generated from: \"\n",
    "        print \"Alignment file: \" + self.bam_file\n",
    "        print \"Motifs file: \" + self.motifs_file\n",
    "        print \"Peaks file: \" + self.peaks_file\n",
    "        print \"TSS annotation file: \" + self.tss_file\n",
    "        #### add other parameters later\n",
    "        \n",
    "    \n",
    "    \n",
    "    def get_motifs(self):\n",
    "        \n",
    "        \n",
    "        \"\"\"\n",
    "        returns BedTool with good motifs (?)\n",
    "        \"\"\"\n",
    "        \n",
    "        \n",
    "        if self.cluster == True:\n",
    "            \n",
    "            motifs_keep_ins = []\n",
    "            motifs_keep_9mer = []\n",
    "            \n",
    "            \n",
    "            # class Motifs to cluster etc\n",
    "            motifs = Motifs(self.bam_file, self.motifs_file)\n",
    "            motifs.read_motifs_file()\n",
    "            motifs.compute_scores_matrices(upstream = upstream_in_matrix, \n",
    "                                           downstream = downstream_in_matrix)\n",
    "            motifs.cluster_motifs()\n",
    "            \n",
    "            # find out which motifs to keep\n",
    "            for regulator in motifs.motif_clusters_insertion_counts.keys():\n",
    "                \n",
    "                keep = motifs.motif_clusters_insertion_counts[regulator]['cluster'] == 'keep'\n",
    "                tmp_motifs_keep = motifs.motif_clusters_insertion_counts[regulator]['motif'][keep] \n",
    "            \n",
    "                # append good motifs\n",
    "                for motif in motifs_keep:\n",
    "                    motifs_keep_ins.append(tuple(motif.split(',')))\n",
    "\n",
    "            # 9mer -- after will keep only one?\n",
    "            #for regulator in motifs.motif_clusters_9mer_counts.keys():\n",
    "                \n",
    "            #    keep = motifs.motif_clusters_9mer_counts[regulator]['cluster'] == 'keep'\n",
    "            #    tmp_motifs_keep = motifs.motif_clusters_9mer_counts[regulator]['motif'][keep] \n",
    "            \n",
    "                # append good motifs\n",
    "            #    for motif in motifs_keep:\n",
    "            #        motifs_keep_9mer.append(tuple(motif.split(',')))\n",
    "\n",
    "            \n",
    "            motifs = pybedtools.BedTool(motifs_keep_ins)\n",
    "            #motifs = pybedtools.BedTool(motifs_keep_9mer)\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            motifs = pybedtools.BedTools(self.motifs_file)\n",
    "        \n",
    "        motifs.sort()\n",
    "        \n",
    "        self.motifs = motifs\n",
    "    \n",
    "    \n",
    "    def get_peaks(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        peaks = pybedtools.Bedtool(self.peaks_file)\n",
    "        peaks = peaks.sort()\n",
    "        \n",
    "        self.peaks = peaks\n",
    "    \n",
    "    \n",
    "    def get_tss(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        tss_genes = pybedtools.BedTool(self.tss_file)\n",
    "        tss_genes = genes_tss.sort()       \n",
    "        \n",
    "        self.tss = tss_genes\n",
    "    \n",
    "    \n",
    "    def assign_motifs_to_genes(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        motifs = self.motifs\n",
    "        peaks = self.peaks\n",
    "        tss = self.tss\n",
    "        \n",
    "        peaks_to_genes = peaks.closest(tss, D = 'b')\n",
    "        motifs_to_genes = motifs.closest(peaks_to_genes)\n",
    "        \n",
    "        return(motifs_to_genes)\n",
    "    \n",
    "    \n",
    "    def make_prior(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        self.get_motifs()\n",
    "        self.get_peaks()\n",
    "        self.get_tss()\n",
    "        \n",
    "        motifs_to_genes = self.assign_motifs_to_genes()\n",
    "        \n",
    "        # format prior output\n",
    "        prior = set()\n",
    "\n",
    "        # format lines in prior\n",
    "        for interaction in motifs_to_genes:\n",
    "            # get fields in string format\n",
    "            interaction = interaction.fields\n",
    "            peak = interaction[9]\n",
    "            # TF name\n",
    "            regulators = interaction[3]\n",
    "            regulators = regulators.split('::')\n",
    "            # target gene\n",
    "            target = interaction[23].split()[1]\n",
    "            target = target.replace('\"', '').replace(';', '')\n",
    "            # distance to feature, if negative it is upstream\n",
    "            distance = int(interaction[24])\n",
    "\n",
    "            for regulator in regulators:\n",
    "                prior_line = (regulator.encode('ascii', 'ignore'),\n",
    "                              target.encode('ascii', 'ignore'),\n",
    "                              peak.encode('ascii', 'ignore'), str(distance))\n",
    "                prior.add(prior_line)\n",
    "                \n",
    "        return(prior)\n",
    "    \n",
    "    \n",
    "    #def write_prior(prior, outfile, outdir):\n",
    "\n",
    "     #   prior_outdir = os.path.join(outdir, 'priors')\n",
    "\n",
    "        #if not os.path.exists(prior_outdir):\n",
    "         #   os.makedirs(prior_outdir)\n",
    "\n",
    "        #filepath = os.path.join(prior_outdir, outfile)\n",
    "\n",
    "        # write prior in table format regulator \\t target\n",
    "        #with open(filepath, 'a') as handle:\n",
    "        #    handle.writelines('\\t'.join(i) + '\\n' for i in prior)\n",
    "\n",
    "        #handle.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "__init__() takes exactly 3 arguments (1 given)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-119-c6ee9390d506>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mMotifs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: __init__() takes exactly 3 arguments (1 given)"
     ]
    }
   ],
   "source": [
    "\n",
    "Motifs()\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('chr10', 10114679, 10114693, 'M5483_1.01', '1.27e-06', '+')]"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# inputs\n",
    "motifs_file = './input-tests/motifs.random100lines.chr10.txt'\n",
    "bam_file = './input-tests/patho_48hrs_2_dedup_mitRM_mapq30p_sorted.bam'\n",
    "\n",
    "example = Motifs(bam_file, motifs_file)\n",
    "example.read_motifs_file()\n",
    "bam_handle = pysam.AlignmentFile(example.bam_file, \"rb\")\n",
    "motif_id = example.motifs.keys()[5]\n",
    "example.motifs.values()[5]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "insertion_counts_matrix = [(9,0,0,0,4,3,5), (9,0,0,0,4,3,5), (0,0,0,0,0,0,2), (0,0,0,0,0,0,2)]\n",
    "rownames_motifs = ['chr10,10114679,10114693,+', 'chr10,10114679,10114694,+', \n",
    "                  'chr10,10114679,10114695,-', 'chr10,10114679,10114696,+']\n",
    "\n",
    "insertion_counts_matrix = pandas.DataFrame(insertion_counts_matrix, index = rownames_motifs)\n",
    "\n",
    "example.insertion_counts = {\"Stat3\": insertion_counts_matrix,\n",
    "                            \"Stat2\": insertion_counts_matrix}\n",
    "\n",
    "example.tn5_9mer_counts = {\"Stat3\": insertion_counts_matrix,\n",
    "                            \"Stat2\": insertion_counts_matrix,}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.6428571428571428"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "insertion_counts_matrix.mean(1).mean()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10,10114679,10114693,+</td>\n",
       "      <td>keep</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10,10114679,10114694,+</td>\n",
       "      <td>keep</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10,10114679,10114695,-</td>\n",
       "      <td>not_keep</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10,10114679,10114696,+</td>\n",
       "      <td>not_keep</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       motif   cluster\n",
       "0  chr10,10114679,10114693,+      keep\n",
       "1  chr10,10114679,10114694,+      keep\n",
       "2  chr10,10114679,10114695,-  not_keep\n",
       "3  chr10,10114679,10114696,+  not_keep"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example.cluster_motifs()\n",
    "example.motif_clusters_insertion_counts['Stat2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('chr10', '10114679', '10114693', '+')\n",
      "('chr10', '10114679', '10114694', '+')\n"
     ]
    }
   ],
   "source": [
    "for i in example.motif_clusters_insertion_counts['Stat2']['motif'][example.motif_clusters_insertion_counts['Stat2']['cluster'] == 'keep']:\n",
    "    print tuple(i.split(','))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114693,+</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114694,+</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114695,-</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114696,+</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           0  1  2  3  4  5  6\n",
       "chr10,10114679,10114693,+  9  0  0  0  4  3  5\n",
       "chr10,10114679,10114694,+  9  0  0  0  4  3  5\n",
       "chr10,10114679,10114695,-  0  0  0  0  0  0  2\n",
       "chr10,10114679,10114696,+  0  0  0  0  0  0  2"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = example.insertion_counts['Stat2'] \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "clust = example.motif_clusters_9mer_counts['Stat2']\n",
    "#clust.query('clust.motif == 0')\n",
    "\n",
    "\n",
    "#df[(df.index in example.motif_clusters_9mer_counts['Stat2']['motif'][example.motif_clusters_9mer_counts['Stat2']['cluster'] == 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114695,-</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114696,+</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           0  1  2  3  4  5  6\n",
       "chr10,10114679,10114695,-  0  0  0  0  0  0  2\n",
       "chr10,10114679,10114696,+  0  0  0  0  0  0  2"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.index.isin(clust['motif'][clust['cluster'] == 0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0)"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrom = 'chr1'\n",
    "start = 20784345\n",
    "end = 20784360\n",
    "Motifs.get_insertions(bam_handle, chrom, start, end)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114693,+</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114694,+</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114695,-</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114696,+</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           0  1  2  3  4  5  6\n",
       "chr10,10114679,10114693,+  9  0  0  0  4  3  5\n",
       "chr10,10114679,10114694,+  9  0  0  0  4  3  5\n",
       "chr10,10114679,10114695,-  0  0  0  0  0  0  2\n",
       "chr10,10114679,10114696,+  0  0  0  0  0  0  2"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "insertion_counts_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "int64\n",
      "[[9 0 0 0 4 3 5]\n",
      " [9 0 0 0 4 3 5]\n",
      " [0 0 0 0 0 0 2]\n",
      " [0 0 0 0 0 0 2]]\n"
     ]
    }
   ],
   "source": [
    "dataset_array = insertion_counts_matrix.values\n",
    "print(dataset_array.dtype)\n",
    "print(dataset_array)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dataset = insertion_counts_matrix\n",
    "\n",
    "# Convert DataFrame to matrix\n",
    "mat = dataset.as_matrix()\n",
    "# Using sklearn\n",
    "km = KMeans(n_clusters = 2)\n",
    "km.fit(mat)\n",
    "# Get cluster assignment labels\n",
    "labels = km.labels_\n",
    "# Format results as a DataFrame\n",
    "results = pandas.DataFrame([dataset.index,labels]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10,10114679,10114693,M5483_1.01</td>\n",
       "      <td>chr10,10114679,10114693,M5483_1.02</td>\n",
       "      <td>chr10,10114679,10114693,M5483_1.03</td>\n",
       "      <td>chr10,10114679,10114693,M5483_1.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    0                                   1  \\\n",
       "0  chr10,10114679,10114693,M5483_1.01  chr10,10114679,10114693,M5483_1.02   \n",
       "1                                   1                                   0   \n",
       "\n",
       "                                    2                                   3  \n",
       "0  chr10,10114679,10114693,M5483_1.03  chr10,10114679,10114693,M5483_1.04  \n",
       "1                                   0                                   0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pandas.DataFrame([dataset.index,labels])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114693,M5483_1.01</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114693,M5483_1.02</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114693,M5483_1.03</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10,10114679,10114693,M5483_1.04</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    0  1  2  3  4  5  6\n",
       "chr10,10114679,10114693,M5483_1.01  9  0  0  0  4  3  5\n",
       "chr10,10114679,10114693,M5483_1.02  1  0  0  4  2  1  0\n",
       "chr10,10114679,10114693,M5483_1.03  0  1  0  0  0  0  2\n",
       "chr10,10114679,10114693,M5483_1.04  0  0  0  0  4  1  5"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([u'chr10,10114679,10114693,M5483_1.01',\n",
       "       u'chr10,10114679,10114693,M5483_1.02',\n",
       "       u'chr10,10114679,10114693,M5483_1.03',\n",
       "       u'chr10,10114679,10114693,M5483_1.04'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10,10114679,10114693,M5483_1.01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10,10114679,10114693,M5483_1.02</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10,10114679,10114693,M5483_1.03</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10,10114679,10114693,M5483_1.04</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    0  1\n",
       "0  chr10,10114679,10114693,M5483_1.01  1\n",
       "1  chr10,10114679,10114693,M5483_1.02  0\n",
       "2  chr10,10114679,10114693,M5483_1.03  0\n",
       "3  chr10,10114679,10114693,M5483_1.04  0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#### GENERATING THE PRIOR...\n",
    "\n",
    "        #\n",
    "        motifs = pybedtools.Bedtool(self.motifs_bed)\n",
    "        motifs = motifs.sort()\n",
    "\n",
    "        #\n",
    "        peaks = pybedtools.Bedtool(self.peaks_bed)\n",
    "        peaks = peaks.sort()\n",
    "\n",
    "        #\n",
    "        genes_tss = pybedtools.BedTool(self.tss_bed)\n",
    "        genes_tss = genes_tss.sort()\n",
    "\n",
    "        # assign peaks to closes feature in feature file\n",
    "        peaks_to_genes = peaks.closest(genes_tss, D = 'b')\n",
    "        motifs_to_genes = motifs.closest(peaks_to_genes)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
