{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pybedtools\n",
    "import pandas\n",
    "from sklearn.cluster import KMeans\n",
    "\n",
    "# inputs\n",
    "motifs_file = './input-tests/motifs.random100lines.chr10.txt'\n",
    "bam_file = './input-tests/patho_48hrs_2_dedup_mitRM_mapq30p_sorted.bam'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "class Motifs:\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    input here should be motifs in peaks already\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    -----------------\n",
    "    Attributes\n",
    "    -----------------\n",
    "    \n",
    "    bam_file: alignment file in BAM format containing reads from an ATAC-seq experiment;\n",
    "              file should be sorted and indexed.\n",
    "    \n",
    "    motifs_file: transcription factor motifs in open regions of the genome formatted as:\n",
    "                 motif-ID chromosome start end strand score p-value q-value matched-sequence\n",
    "                 tab delimited, no header, should include only motifs in peaks\n",
    "    \n",
    "    motifs: hold a dictionary with motifs from motifs_file;\n",
    "            ---\n",
    "            {TF1:[(occurence1_chrom, occurence1_start, occurence1_end, occurence1_p-value, occurence1_strand),\n",
    "                  (occurence2_chrom, occurence2_start, occurence2_end, occurence2_p-value, occurence2_strand)],\n",
    "             TF2:[(occurence1_chrom, occurence1_start, occurence1_end, occurence1_p-value, occurence1_strand)]...}        \n",
    "            ---\n",
    "            Keys: motif IDs -- will be used later to build the prior, thus I use transcription factors as IDs.\n",
    "            Values: list of occurences in the genome of the motif(s) associated with identifier.\n",
    "\n",
    "    insertion_counts: \n",
    "    \n",
    "    tn5_9mer_counts:\n",
    "    \n",
    "    motif_clusters_insertion_counts:\n",
    "\n",
    "    motif_clusters_9mer_counts:\n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, bam_file, motifs_file):\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        self.bam_file = bam_file\n",
    "        self.motifs_file = motifs_file \n",
    "        self.motifs = {}\n",
    "        self.insertion_counts = {}\n",
    "        self.tn5_9mer_counts = {}\n",
    "        self.motif_clusters_insertion_counts = {}\n",
    "        self.motif_clusters_9mer_counts = {}\n",
    "        #self.cluster = True #False\n",
    "        #self.motif_clusters = []\n",
    "        \n",
    "    def parse_motifs_file(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        from motifs_file to self.motifs\n",
    "        \"\"\"\n",
    "        \n",
    "        motifs_file = self.motifs_file\n",
    "\n",
    "        with open(motifs_file, \"r\") as motifs_handle:\n",
    "\n",
    "            for motif_occurence in motifs_handle:\n",
    "\n",
    "                motif_occurence = motif_occurence.split()\n",
    "\n",
    "                motif_id = motif_occurence[0]\n",
    "                chrom = motif_occurence[1]\n",
    "                start = motif_occurence[2]\n",
    "                end = motif_occurence[3]\n",
    "                strand = motif_occurence[4]\n",
    "                score = motif_occurence[6] # p-value\n",
    "\n",
    "                to_bed = [chrom, int(start) - 1, int(end) - 1, motif_id, score, strand]\n",
    "\n",
    "                if motif_id in self.motifs:\n",
    "                    self.motifs[motif_id].append(tuple(to_bed))\n",
    "                else:\n",
    "                    self.motifs[motif_id] = [tuple(to_bed)]\n",
    "    \n",
    "\n",
    "    @staticmethod \n",
    "    def get_insertions(bam_handle, chrom, start, end, upstream = 50, downstream = 50):\n",
    "        \n",
    "        \"\"\"\n",
    "        returns number of \n",
    "        \"\"\"\n",
    "        \n",
    "\n",
    "        # initialize counts vector at 0 for all positions\n",
    "        region_length = (end - start) + 1\n",
    "        insertion_counts = [0]*region_length\n",
    "\n",
    "        # fetch reads mapping to the region specified as input\n",
    "        reads = bam_handle.fetch(chrom, start, end)\n",
    "\n",
    "        # each read represents a potential insertion within region\n",
    "        for read in reads:\n",
    "            \n",
    "            if read.is_reverse: \n",
    "                # offset by 5 bp\n",
    "                insertion_pos = read.reference_end - 5 \n",
    "            else:\n",
    "                # offset by 4 bp\n",
    "                insertion_pos = read.reference_start + 4\n",
    "            \n",
    "            pos = insertion_pos - start \n",
    "            \n",
    "            # make sure pos is within region\n",
    "            if pos in range(0, region_length):\n",
    "                insertion_counts[pos] += 1  \n",
    "                    \n",
    "        return tuple(insertion_counts)\n",
    "        \n",
    "    @staticmethod\n",
    "    def get_tn5_9mer(insertion_counts, up_offset = 4, down_offset = 5):\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        \"\"\"\n",
    "    \n",
    "        tn5_9mer_counts = list(insertion_counts)\n",
    "\n",
    "        region = range(0, len(insertion_counts))\n",
    "\n",
    "        for pos in region:\n",
    "            for idx in range(pos - up_offset, pos + down_offset + 1):\n",
    "                if idx in region and idx != pos:\n",
    "                    tn5_9mer_counts[idx] += insertion_counts[pos]\n",
    "        \n",
    "        return tuple(tn5_9mer_counts)\n",
    "            \n",
    "    \n",
    "    def compute_scores_matrices(self, upstream = 50, downstream = 50):\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        self.parse_motifs_file()\n",
    "        \n",
    "        bam_handle = pysam.AlignmentFile(self.bam_file, \"rb\")\n",
    "        \n",
    "        for motif_id in self.motifs.keys(): ## Paralelize\n",
    "            \n",
    "            insertion_counts_mat = []\n",
    "            tn5_9mer_counts_mat = []\n",
    "            rownames_motifs = []\n",
    "            \n",
    "            motifs = pybedtools.Bedtool(self.motifs[motif_id])\n",
    "            motifs = motifs.sort()\n",
    "            \n",
    "            for motif in motifs:\n",
    "                \n",
    "                # get motif coordinates -- FIMO output is a closed interval, so add 1 to end (python)\n",
    "                chrom = motif.chrom\n",
    "                center = round(stat.median(range(motif.start, motif.end + 1)))\n",
    "                start = center - upstream\n",
    "                end = center + downstream\n",
    "                \n",
    "                insertion_counts = get_insertions(bam_handle, chrom, start, end)\n",
    "                insertion_counts_mat.append(insertion_counts)\n",
    "                \n",
    "                tn5_9mer_counts = get_tn5_9mer(insertion_counts)\n",
    "                tn5_9mer_counts_mat.append(tn5_9mer_counts)\n",
    "                \n",
    "                motif_holder = \"_\".join([motif.chrom, str(motif.start), str(motif.end), motif.name])\n",
    "                rownames_motifs.append(motif_holder)\n",
    "                \n",
    "            \n",
    "            insertion_counts_df = pandas.DataFrame(insertion_counts_mat, \n",
    "                                                   index = rownames_motifs)\n",
    "            \n",
    "            tn5_9mer_counts_df = pandas.DataFrame(tn5_9mer_counts_mat, \n",
    "                                                  index = rownames_motifs)\n",
    "            \n",
    "            self.insertion_counts[motif_id] = insertion_counts_df\n",
    "            self.tn5_9mer_counts[motif_id] = tn5_9mer_counts_df\n",
    "            \n",
    "\n",
    "    def cluster_motifs(self):\n",
    "        \n",
    "        ## Paralelize, as of now I am going to do both ins and 9mer (choose one later?)\n",
    "        \n",
    "        # insertion counts\n",
    "        for motif_holder in self.insertion_counts.keys:\n",
    "            \n",
    "            insertion_counts_df = self.insertion_counts[motif_id]\n",
    "            insertion_counts_mat = insertion_counts_df.as_matrix()\n",
    "\n",
    "            motif_clusters = KMeans(n_clusters = 2)\n",
    "            motif_clusters.fit(insertion_counts_mat)\n",
    "            labels = motif_clusters.labels_\n",
    "            results = pandas.DataFrame([insertion_counts_df.index, labels]).T\n",
    "\n",
    "            self.motif_clusters_insertion_counts[motif_holder] = results\n",
    "            \n",
    "            \n",
    "        # 9mers\n",
    "        for motif_holder in self.tn5_9mer_counts.keys:\n",
    "            \n",
    "            tn5_9mer_counts_df = self.tn5_9mer_counts[motif_id]\n",
    "\n",
    "            tn5_9mer_counts_mat = tn5_9mer_counts_df.as_matrix()\n",
    "\n",
    "            motif_clusters = KMeans(n_clusters = 2)\n",
    "            motif_clusters.fit(tn5_9mer_counts_mat)\n",
    "            labels = motif_clusters.labels_\n",
    "            results = pandas.DataFrame([tn5_9mer_counts_df.index, labels]).T\n",
    "            \n",
    "            self.motif_clusters_9mer_counts[motif_holder] = results\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "example = Motifs(bam_file, motifs_file)\n",
    "example.parse_motifs_file()\n",
    "bam_handle = pysam.AlignmentFile(example.bam_file, \"rb\")\n",
    "motif_id = example.motifs.keys()[5]\n",
    "chrom = 'chr1'\n",
    "start = 20784345\n",
    "end = 20784360\n",
    "Motifs.get_insertions(bam_handle, chrom, start, end)\n",
    "\n",
    "\n",
    "insertion_counts_matrix = [(9,0,0,0,4,3,5), (1,0,0,4,2,1,0), (0,1,0,0,0,0,2), (0,0,0,0,4,1,5)]\n",
    "rownames_motifs = ['chr10_10114679_10114693_M5483_1.01', 'chr10_10114679_10114693_M5483_1.02', \n",
    "                  'chr10_10114679_10114693_M5483_1.03', 'chr10_10114679_10114693_M5483_1.04']\n",
    "\n",
    "insertion_counts_matrix = pandas.DataFrame(insertion_counts_matrix, index = rownames_motifs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10_10114679_10114693_M5483_1.01</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_10114679_10114693_M5483_1.02</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_10114679_10114693_M5483_1.03</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_10114679_10114693_M5483_1.04</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    0  1  2  3  4  5  6\n",
       "chr10_10114679_10114693_M5483_1.01  9  0  0  0  4  3  5\n",
       "chr10_10114679_10114693_M5483_1.02  1  0  0  4  2  1  0\n",
       "chr10_10114679_10114693_M5483_1.03  0  1  0  0  0  0  2\n",
       "chr10_10114679_10114693_M5483_1.04  0  0  0  0  4  1  5"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "insertion_counts_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "int64\n",
      "[[9 0 0 0 4 3 5]\n",
      " [1 0 0 4 2 1 0]\n",
      " [0 1 0 0 0 0 2]\n",
      " [0 0 0 0 4 1 5]]\n"
     ]
    }
   ],
   "source": [
    "dataset_array = insertion_counts_matrix.values\n",
    "print(dataset_array.dtype)\n",
    "print(dataset_array)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dataset = insertion_counts_matrix\n",
    "\n",
    "# Convert DataFrame to matrix\n",
    "mat = dataset.as_matrix()\n",
    "# Using sklearn\n",
    "km = KMeans(n_clusters = 3)\n",
    "km.fit(mat)\n",
    "# Get cluster assignment labels\n",
    "labels = km.labels_\n",
    "# Format results as a DataFrame\n",
    "results = pandas.DataFrame([dataset.index,labels]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10_10114679_10114693_M5483_1.01</td>\n",
       "      <td>chr10_10114679_10114693_M5483_1.02</td>\n",
       "      <td>chr10_10114679_10114693_M5483_1.03</td>\n",
       "      <td>chr10_10114679_10114693_M5483_1.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    0                                   1  \\\n",
       "0  chr10_10114679_10114693_M5483_1.01  chr10_10114679_10114693_M5483_1.02   \n",
       "1                                   0                                   1   \n",
       "\n",
       "                                    2                                   3  \n",
       "0  chr10_10114679_10114693_M5483_1.03  chr10_10114679_10114693_M5483_1.04  \n",
       "1                                   1                                   2  "
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pandas.DataFrame([dataset.index,labels])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10_10114679_10114693_M5483_1.01</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_10114679_10114693_M5483_1.02</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_10114679_10114693_M5483_1.03</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_10114679_10114693_M5483_1.04</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    0  1  2  3  4  5  6\n",
       "chr10_10114679_10114693_M5483_1.01  9  0  0  0  4  3  5\n",
       "chr10_10114679_10114693_M5483_1.02  1  0  0  4  2  1  0\n",
       "chr10_10114679_10114693_M5483_1.03  0  1  0  0  0  0  2\n",
       "chr10_10114679_10114693_M5483_1.04  0  0  0  0  4  1  5"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([u'chr10_10114679_10114693_M5483_1.01',\n",
       "       u'chr10_10114679_10114693_M5483_1.02',\n",
       "       u'chr10_10114679_10114693_M5483_1.03',\n",
       "       u'chr10_10114679_10114693_M5483_1.04'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#### GENERATING THE PRIOR...\n",
    "\n",
    "        #\n",
    "        motifs = pybedtools.Bedtool(self.motifs_bed)\n",
    "        motifs = motifs.sort()\n",
    "\n",
    "        #\n",
    "        peaks = pybedtools.Bedtool(self.peaks_bed)\n",
    "        peaks = peaks.sort()\n",
    "\n",
    "        #\n",
    "        genes_tss = pybedtools.BedTool(self.tss_bed)\n",
    "        genes_tss = genes_tss.sort()\n",
    "\n",
    "        # assign peaks to closes feature in feature file\n",
    "        peaks_to_genes = peaks.closest(genes_tss, D = 'b')\n",
    "        motifs_to_genes = motifs.closest(peaks_to_genes)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
