{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "import pysam\n",
    "import pybedtools\n",
    "import pandas\n",
    "import statistics as stat\n",
    "from sklearn.cluster import KMeans\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "class Motifs:\n",
    "\n",
    "    \"\"\"\n",
    "    -----------------\n",
    "    Attributes\n",
    "    -----------------\n",
    "    \n",
    "    bam_file: alignment file in BAM format containing reads from an ATAC-seq experiment;\n",
    "              file should be sorted and indexed.\n",
    "    \n",
    "    motifs_file: transcription factor motifs in open regions of the genome formatted as:\n",
    "                 chromosome start end motif-ID score strand (score can be e-value, p-value, doesn't matter...)\n",
    "                 tab delimited, no header, should include only motifs in peaks\n",
    "    \n",
    "    motifs: holds a dictionary with motifs from motifs_file;\n",
    "            ---\n",
    "            {TF1:[(occurence1_chrom, occurence1_start, occurence1_end, occurence1_p-value, occurence1_strand),\n",
    "                  (occurence2_chrom, occurence2_start, occurence2_end, occurence2_p-value, occurence2_strand)],\n",
    "             TF2:[(occurence1_chrom, occurence1_start, occurence1_end, occurence1_p-value, occurence1_strand)]...}        \n",
    "            ---\n",
    "            Keys: motif IDs -- will be used later to build the prior, thus I use transcription factors as IDs.\n",
    "            Values: list of occurences in the genome of the motif(s) associated with identifier.\n",
    "\n",
    "    insertion_counts: holds a dictionary with matrices of insertion counts around motifs for a particular transcription factor\n",
    "                      ---\n",
    "                      Keys: motif holders -- will be used later to build the prior \n",
    "                      Values: pandas DataFrame with insertion counts around each motif for the motif(s) associated with identifier,\n",
    "                              rownames are motif holders -- chrom,start,end,tf,score,strand\n",
    "\n",
    "    tn5_9mer_counts: holds a dictionary with matrices of 9mer counts around motifs for a particular transcription factor\n",
    "                     ---\n",
    "                     Keys: motif holders -- will be used later to build the prior\n",
    "                     Values: pandas DataFrame with 9mer counts around each motif for the motif(s) associated with identifier,\n",
    "                             rownames are motif holders as above\n",
    "\n",
    "    motif_clusters_insertion_counts: hold a dictionary with clusters of motifs to 'keep' or 'not_keep' -- generated using insertion counts\n",
    "                                     Keys: motif IDs (as above)\n",
    "                                     Values: pandas DataFrame with clusters of motifs associated with identifier.\n",
    "                                     rownames are motif holders as above\n",
    "\n",
    "    motif_clusters_9mer_counts: hold a dictionary with clusters of motifs to 'keep' or 'not_keep' -- generated using 9mer counts\n",
    "                                     Keys: motif IDs (as above)\n",
    "                                     Values: pandas DataFrame with clusters of motifs associated with identifier.\n",
    "                                     rownames are motif holders as above\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, bam_file, motifs_file):\n",
    "        \n",
    "        self.bam_file = bam_file\n",
    "        self.motifs_file = motifs_file \n",
    "        self.motifs = {}\n",
    "        self.insertion_counts = {}\n",
    "        self.tn5_9mer_counts = {}\n",
    "        self.motif_clusters_insertion_counts = {}\n",
    "        self.motif_clusters_9mer_counts = {}\n",
    "        \n",
    "    def read_motifs_file(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        Reads in a motifs file and saves a dictionary of motifs for each transcription factor in self.motifs \n",
    "        \"\"\"\n",
    "                \n",
    "        motifs_file = self.motifs_file\n",
    "\n",
    "        with open(motifs_file, \"r\") as motifs_handle:\n",
    "\n",
    "            for motif_occurence in motifs_handle:\n",
    "\n",
    "                motif_occurence = motif_occurence.split()\n",
    "\n",
    "                motif_id = motif_occurence[3]\n",
    "                chrom = motif_occurence[0]\n",
    "                start = motif_occurence[1]\n",
    "                end = motif_occurence[2]\n",
    "                strand = motif_occurence[5]\n",
    "                score = motif_occurence[4]\n",
    "\n",
    "                to_bed = [chrom, int(start), int(end), motif_id, score, strand]\n",
    "\n",
    "                if motif_id in self.motifs:\n",
    "                    self.motifs[motif_id].append(tuple(to_bed))\n",
    "                else:\n",
    "                    self.motifs[motif_id] = [tuple(to_bed)]\n",
    "    \n",
    "\n",
    "    def get_insertions(self, chrom, start, end, upstream = 50, downstream = 50):\n",
    "        \n",
    "        \"\"\"\n",
    "        Returns insertion counts at a given region of the genome\n",
    "        \"\"\"\n",
    "        \n",
    "        bam_handle = pysam.AlignmentFile(self.bam_file, \"rb\")\n",
    "        \n",
    "        # initialize counts vector at 0 for all positions\n",
    "        region_length = int((end - start) + 1)\n",
    "        insertion_counts = [0]*region_length\n",
    "\n",
    "        # fetch reads mapping to the region specified as input\n",
    "        reads = bam_handle.fetch(chrom, start, end)\n",
    "\n",
    "        # each read represents a potential insertion within region\n",
    "        for read in reads:\n",
    "            \n",
    "            if read.is_reverse: \n",
    "                # offset by 5 bp\n",
    "                insertion_pos = read.reference_end - 5 \n",
    "            else:\n",
    "                # offset by 4 bp\n",
    "                insertion_pos = read.reference_start + 4\n",
    "            \n",
    "            pos = insertion_pos - start \n",
    "            \n",
    "            # make sure pos is within region\n",
    "            if pos in range(0, region_length):\n",
    "                insertion_counts[pos] += 1  \n",
    "                    \n",
    "        return tuple(insertion_counts)\n",
    "        \n",
    "    @staticmethod\n",
    "    def get_tn5_9mer(insertion_counts, up_offset = 4, down_offset = 5):\n",
    "        \n",
    "        \"\"\"\n",
    "        Smooth insertion counts to Tn5 occupancy (9mer track)\n",
    "        \"\"\"\n",
    "    \n",
    "        tn5_9mer_counts = list(insertion_counts)\n",
    "\n",
    "        region = range(0, len(insertion_counts))\n",
    "\n",
    "        for pos in region:\n",
    "            for idx in range(pos - up_offset, pos + down_offset + 1):\n",
    "                if idx in region and idx != pos:\n",
    "                    tn5_9mer_counts[idx] += insertion_counts[pos]\n",
    "        \n",
    "        return tuple(tn5_9mer_counts)\n",
    "            \n",
    "    \n",
    "    def compute_scores_matrices(self, upstream, downstream):\n",
    "        \n",
    "        \"\"\"\n",
    "        Generate matrices of insertion and 9mer counts for all motifs in self.motifs\n",
    "        \"\"\"\n",
    "        \n",
    "        \n",
    "        for motif_id in self.motifs.keys(): ## Paralelize\n",
    "            \n",
    "            insertion_counts_mat = []\n",
    "            tn5_9mer_counts_mat = []\n",
    "            rownames_motifs = []\n",
    "            \n",
    "            motifs = pybedtools.BedTool(self.motifs[motif_id])\n",
    "            motifs = motifs.sort()\n",
    "            \n",
    "            for motif in motifs:\n",
    "                \n",
    "                # get motif coordinates -- FIMO output is a closed interval, so add 1 to end (python)\n",
    "                chrom = str(motif.chrom)\n",
    "                center = round(stat.median(range(motif.start, motif.end + 1)))\n",
    "                start = int(center - upstream)\n",
    "                end = int(center + downstream)\n",
    "                \n",
    "                insertion_counts = self.get_insertions(chrom, start, end)\n",
    "                insertion_counts_mat.append(insertion_counts)\n",
    "                \n",
    "                tn5_9mer_counts = Motifs.get_tn5_9mer(insertion_counts)\n",
    "                tn5_9mer_counts_mat.append(tn5_9mer_counts)\n",
    "                \n",
    "                motif_holder = \",\".join([motif.chrom, str(motif.start), str(motif.end), motif.name, motif.score, motif.strand])\n",
    "                rownames_motifs.append(motif_holder)\n",
    "                \n",
    "            \n",
    "            insertion_counts_df = pandas.DataFrame(insertion_counts_mat, \n",
    "                                                   index = rownames_motifs)\n",
    "            \n",
    "            tn5_9mer_counts_df = pandas.DataFrame(tn5_9mer_counts_mat, \n",
    "                                                  index = rownames_motifs)\n",
    "            \n",
    "            self.insertion_counts[motif_id] = insertion_counts_df\n",
    "            self.tn5_9mer_counts[motif_id] = tn5_9mer_counts_df\n",
    "            \n",
    "\n",
    "    def cluster_motifs(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        Use K-Means to cluster motifs for a given transcription factor in two clusters for all TFs in self.motifs\n",
    "        \"\"\"\n",
    "        ## Paralelize, as of now I am going to do both ins and 9mer (choose one later?)\n",
    "        \n",
    "        # insertion counts\n",
    "        for motif_id in self.insertion_counts.keys():\n",
    "            \n",
    "            insertion_counts_df = self.insertion_counts[motif_id]\n",
    "            insertion_counts_mat = insertion_counts_df.as_matrix()\n",
    "\n",
    "            motif_clusters = KMeans(n_clusters = 2)\n",
    "            motif_clusters.fit(insertion_counts_mat)\n",
    "            labels = motif_clusters.labels_\n",
    "            results = pandas.DataFrame([insertion_counts_df.index, labels]).T\n",
    "            results.columns = ['motif', 'cluster']     \n",
    "            \n",
    "            clust0 = insertion_counts_df[insertion_counts_df.index.isin(results['motif'][results['cluster'] == 0])]\n",
    "            clust1 = insertion_counts_df[insertion_counts_df.index.isin(results['motif'][results['cluster'] == 1])]\n",
    "            \n",
    "            clust0_mean = clust0.mean(1).mean()\n",
    "            clust1_mean = clust1.mean(1).mean()\n",
    "            \n",
    "            if clust0_mean > clust1_mean:\n",
    "                \n",
    "                results['cluster'][results['cluster'] == 0] = 'keep'\n",
    "                results['cluster'][results['cluster'] == 1] = 'not_keep'\n",
    "            \n",
    "            if clust1_mean > clust0_mean:\n",
    "                \n",
    "                results['cluster'][results['cluster'] == 1] = 'keep'\n",
    "                results['cluster'][results['cluster'] == 0] = 'not_keep'\n",
    "            \n",
    "            self.motif_clusters_insertion_counts[motif_id] = results\n",
    "            \n",
    "            \n",
    "            \n",
    "        # 9mers\n",
    "        for motif_id in self.tn5_9mer_counts.keys():\n",
    "            \n",
    "            tn5_9mer_counts_df = self.tn5_9mer_counts[motif_id]\n",
    "            tn5_9mer_counts_mat = tn5_9mer_counts_df.as_matrix()\n",
    "\n",
    "            motif_clusters = KMeans(n_clusters = 2)\n",
    "            motif_clusters.fit(tn5_9mer_counts_mat)\n",
    "            labels = motif_clusters.labels_\n",
    "            results = pandas.DataFrame([tn5_9mer_counts_df.index, labels]).T\n",
    "            results.columns = ['motif', 'cluster']\n",
    "            \n",
    "            clust0 = insertion_counts_df[insertion_counts_df.index.isin(results['motif'][results['cluster'] == 0])]\n",
    "            clust1 = insertion_counts_df[insertion_counts_df.index.isin(results['motif'][results['cluster'] == 1])]\n",
    "            \n",
    "            clust0_mean = clust0.mean(1).mean()\n",
    "            clust1_mean = clust1.mean(1).mean()\n",
    "            \n",
    "            if clust0_mean > clust1_mean:\n",
    "                \n",
    "                results['cluster'][results['cluster'] == 0] = 'keep'\n",
    "                results['cluster'][results['cluster'] == 1] = 'not_keep'\n",
    "            \n",
    "            if clust1_mean > clust0_mean:\n",
    "                \n",
    "                results['cluster'][results['cluster'] == 1] = 'keep'\n",
    "                results['cluster'][results['cluster'] == 0] = 'not_keep'\n",
    "            \n",
    "            self.motif_clusters_9mer_counts[motif_id] = results\n",
    "\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "class Edge:\n",
    "    \n",
    "    \"\"\"\n",
    "    Container for edges in the prior\n",
    "    ----------\n",
    "    Attributes\n",
    "    ----------\n",
    "    regulator: transcription factor name\n",
    "    target: target gene name\n",
    "    weight: weight to be put in the prior matrix\n",
    "    motif: motif coordinates\n",
    "    gene: gene coordinates -- obs. NOW I HAVE THE TSS HERE, SHOULD BE CHANGED\n",
    "    distance: peak distance to gene \n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, regulator, target, weight, motif, gene, distance):\n",
    "        \n",
    "        self.regulator = regulator\n",
    "        self.target = target\n",
    "        self.weight = weight\n",
    "        self.motif = motif\n",
    "        self.gene = gene\n",
    "        self.distance = distance\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "class Prior:\n",
    "    \n",
    "    \"\"\"\n",
    "    ----------------\n",
    "    Attributes\n",
    "    -----------------\n",
    "    \n",
    "    bam_file: aligned ATAC-seq BAM file\n",
    "    \n",
    "    motifs_file: motifs annotation in FIMO format IN PEAKS\n",
    "    \n",
    "    peaks_file: ATAC peaks in BED format (e.g. called by MACS)\n",
    "    \n",
    "    tss_file: TSS annotation in BED format\n",
    "    \n",
    "    transcription_factors:\n",
    "    \n",
    "    target_genes:\n",
    "    \n",
    "    upstream_in_matrix: \n",
    "    \n",
    "    downstream_in_matrix:\n",
    "    \n",
    "    cluster: whether to cluster motifs or not \n",
    "    \n",
    "    #,max_upstream_distance = float(\"Inf\"), max_downstream_distance = float(\"Inf\")):\n",
    "    \n",
    "    max_upstream_distance: bp upstream allowed for assignment peak to gene\n",
    "    \n",
    "    max_downstream_distance: bp downstream allowed for assignment peak to gene\n",
    "        \n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, bam_file, motifs_file,  peaks_file, tss_file,\n",
    "                 transcription_factors, target_genes, cluster = True,\n",
    "                 upstream_in_matrix = 50, downstream_in_matrix = 50):\n",
    "        \n",
    "        self.bam_file = bam_file\n",
    "        self.motifs_file = motifs_file \n",
    "        self.peaks_file = peaks_file   \n",
    "        self.tss_file = tss_file\n",
    "        self.transcription_factors = transcription_factors\n",
    "        self.target_genes = target_genes\n",
    "        self.upstream_in_matrix = upstream_in_matrix\n",
    "        self.downstream_in_matrix = downstream_in_matrix\n",
    "        self.cluster = cluster \n",
    "        #self.max_upstream_distance = max_upstream_distance     \n",
    "        #self.max_downstream_distance = max_downstream_distance \n",
    "        \n",
    "        self.motifs = []\n",
    "        self.peaks = []\n",
    "        self.tss = []\n",
    "        self.edges = []\n",
    "        \n",
    "        \n",
    "    def __str__(self):\n",
    "        \n",
    "        \n",
    "        print \"Prior generated from: \"\n",
    "        print \"Alignment file: \" + self.bam_file\n",
    "        print \"Motifs file: \" + self.motifs_file\n",
    "        print \"Peaks file: \" + self.peaks_file\n",
    "        print \"TSS annotation file: \" + self.tss_file\n",
    "        #### add other parameters later\n",
    "        \n",
    "    \n",
    "    \n",
    "    def get_motifs(self):\n",
    "        \n",
    "        \n",
    "        \"\"\"\n",
    "        returns a BedTool with good motifs: all if no clustering, else motifs in cluster 'keep'.\n",
    "        \"\"\"\n",
    "        \n",
    "        if self.cluster == True:\n",
    "            \n",
    "            motifs_keep_ins = []\n",
    "            motifs_keep_9mer = []\n",
    "            \n",
    "            \n",
    "            # class Motifs to cluster etc\n",
    "            motifs = Motifs(self.bam_file, self.motifs_file)\n",
    "            motifs.read_motifs_file()\n",
    "            motifs.compute_scores_matrices(upstream = self.upstream_in_matrix, \n",
    "                                           downstream = self.downstream_in_matrix)\n",
    "            motifs.cluster_motifs()\n",
    "            \n",
    "            # find out which motifs to keep\n",
    "            for regulator in motifs.motif_clusters_insertion_counts.keys():\n",
    "                \n",
    "                keep = motifs.motif_clusters_insertion_counts[regulator]['cluster'] == 'keep'\n",
    "                tmp_motifs_keep = motifs.motif_clusters_insertion_counts[regulator]['motif'][keep] \n",
    "            \n",
    "                # append good motifs\n",
    "                for motif in tmp_motifs_keep:\n",
    "                    motifs_keep_ins.append(tuple(motif.split(',')))\n",
    "\n",
    "            # 9mer -- after will keep only one?\n",
    "            #for regulator in motifs.motif_clusters_9mer_counts.keys():\n",
    "                \n",
    "            #    keep = motifs.motif_clusters_9mer_counts[regulator]['cluster'] == 'keep'\n",
    "            #    tmp_motifs_keep = motifs.motif_clusters_9mer_counts[regulator]['motif'][keep] \n",
    "            \n",
    "                # append good motifs\n",
    "            #    for motif in motifs_keep:\n",
    "            #        motifs_keep_9mer.append(tuple(motif.split(',')))\n",
    "\n",
    "            \n",
    "            motifs = pybedtools.BedTool(motifs_keep_ins)\n",
    "            #motifs = pybedtools.BedTool(motifs_keep_9mer)\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            motifs = pybedtools.BedTool(self.motifs_file)\n",
    "        \n",
    "        self.motifs = motifs.sort()\n",
    "    \n",
    "    \n",
    "    def get_peaks(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        returns a BedTool with peaks in peaks_file\n",
    "        \"\"\"\n",
    "        \n",
    "        peaks = pybedtools.BedTool(self.peaks_file)\n",
    "        peaks = peaks.sort()\n",
    "        \n",
    "        self.peaks = peaks\n",
    "    \n",
    "    \n",
    "    def get_tss(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        returns a BedTool with TSS coordinates in tss_file\n",
    "        \"\"\"\n",
    "        \n",
    "        tss_genes = pybedtools.BedTool(self.tss_file)\n",
    "        tss_genes = tss_genes.sort()       \n",
    "        \n",
    "        self.tss = tss_genes\n",
    "    \n",
    "    \n",
    "    def assign_motifs_to_genes(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        assign motifs to genes -- first assign peaks to genes and then all motifs in that peak to that gene.\n",
    "        returns a BedTool with motif - peak - gene coordinates\n",
    "        \"\"\"\n",
    "        \n",
    "        motifs = self.motifs        \n",
    "        peaks = self.peaks\n",
    "        tss = self.tss\n",
    "        \n",
    "        peaks_to_genes = peaks.closest(tss, D = 'b')\n",
    "        motifs_to_genes = motifs.closest(peaks_to_genes.sort())\n",
    "                \n",
    "        return(motifs_to_genes)\n",
    "    \n",
    "    \n",
    "    def get_edges(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        motifs_to_genes = self.assign_motifs_to_genes()\n",
    "\n",
    "        for assignment in motifs_to_genes:\n",
    "            \n",
    "            assignment = assignment.fields\n",
    "            regulators = assignment[3].split('::')\n",
    "            target = assignment[18].split()[0]\n",
    "            distance = int(assignment[21])\n",
    "            weight = 1\n",
    "            \n",
    "            motif = (assignment[0].encode('ascii','ignore'), \n",
    "                     assignment[1].encode('ascii','ignore'), \n",
    "                     assignment[2].encode('ascii','ignore'), \n",
    "                     assignment[3].encode('ascii','ignore'))\n",
    "            \n",
    "            gene = (assignment[15].encode('ascii','ignore'), \n",
    "                    assignment[16].encode('ascii','ignore'), \n",
    "                    assignment[17].encode('ascii','ignore'), \n",
    "                    assignment[18].encode('ascii','ignore'))\n",
    "            \n",
    "            for regulator in regulators:\n",
    "                \n",
    "                edge = Edge(regulator.encode('ascii','ignore'), \n",
    "                            target.encode('ascii','ignore'), \n",
    "                            weight,\n",
    "                            motif, \n",
    "                            gene, \n",
    "                            distance)\n",
    "                \n",
    "                self.edges.append(edge)\n",
    "        \n",
    "        \n",
    "    def make_prior(self):\n",
    "            \n",
    "        \"\"\"\n",
    "        call all functions to generate the prior and fill in edges attribute\n",
    "        \"\"\"\n",
    "            \n",
    "        self.get_motifs()\n",
    "        self.get_peaks()\n",
    "        self.get_tss()\n",
    "        self.get_edges()    \n",
    "            \n",
    "            \n",
    "    def make_prior_matrix(self):\n",
    "            \n",
    "        \"\"\"\n",
    "        returns a pandas DataFrame with the prior matrix using all edges in self.edges\n",
    "        rows are genes in self.target_genes, columns are the TFs in self.transcription_factors\n",
    "        \"\"\"\n",
    "            \n",
    "        prior = pandas.DataFrame(0, index = self.target_genes, \n",
    "                                    columns = self.transcription_factors)\n",
    "            \n",
    "        for edge in self.edges:\n",
    "            \n",
    "            if edge.target in self.target_genes and edge.regulator in self.transcription_factors:\n",
    "\n",
    "                prior.ix[edge.target, edge.regulator] = edge.weight\n",
    "        \n",
    "        return prior\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# inputs\n",
    "motifs_file = './input-tests/motifs_test_file.bed'\n",
    "bam_file = './input-tests/bam_test_file.bam'\n",
    "peaks_file = './input-tests/peaks_test_file.bed'\n",
    "tss_file = './input-tests/tsses_test_file.bed'\n",
    "transcription_factors = ['TF1', 'TF2', 'TF3']\n",
    "target_genes = ['gene1', 'gene2']\n",
    "\n",
    "prior = Prior(bam_file, motifs_file,  peaks_file, tss_file,\n",
    "              transcription_factors, target_genes, cluster = True,\n",
    "              upstream_in_matrix = 50, downstream_in_matrix = 50)\n",
    "prior.make_prior()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(prior.edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TF2 regulates gene2\n",
      "TF1 regulates gene2\n",
      "TF3 regulates gene1\n"
     ]
    }
   ],
   "source": [
    "for x in prior.edges:\n",
    "    print x.regulator + ' regulates ' + x.target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF1</th>\n",
       "      <th>TF2</th>\n",
       "      <th>TF3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       TF1  TF2  TF3\n",
       "gene1    0    0    1\n",
       "gene2    1    1    0"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prior.make_prior_matrix()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('chr1', '20725699', '20725699', 'gene2')"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prior.edges[0].gene # FIX?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('chr1', '20725795', '20725810', 'TF2')"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prior.edges[0].motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF1</th>\n",
       "      <th>TF2</th>\n",
       "      <th>TF3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       TF1  TF2  TF3\n",
       "gene1    1    0    1\n",
       "gene2    1    1    0"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prior = Prior(bam_file, motifs_file,  peaks_file, tss_file,\n",
    "              transcription_factors, target_genes, cluster = False,\n",
    "              upstream_in_matrix = 50, downstream_in_matrix = 50)\n",
    "prior.make_prior()\n",
    "prior.make_prior_matrix()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Obs.\n",
    "should I add the insertion/9mer counts at motif to the Edge attribute?\n",
    "how to do with the genes? HTSeq count?\n",
    "cache options here might be important...\n",
    "suggestions? :)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
